src/aleph/web/controllers/storage.py

```diff
@@ -56,20 +92,133 @@ async def add_storage_json_controller(request: web.Request):
 
 
+async def _verify_message_signature(
+    pending_message: BasePendingMessage, chain_service: ChainService
+) -> None:
+    try:
+        await chain_service.verify_signature(pending_message)
+    except InvalidSignature:
+        raise web.HTTPForbidden()
+
+
+async def _verify_user_balance(
+    pending_message_db: PendingMessageDb, session: DbSession, size: int
+) -> None:
+    current_balance = get_total_balance(
+        session=session, address=pending_message_db.sender
+    ) or Decimal(0)
+    required_balance = (size / MiB) / 3
+    current_cost_for_user = get_total_cost_for_address(
+        session=session, address=pending_message_db.sender
+    )
+    if current_balance < (Decimal(required_balance) + current_cost_for_user):
+        raise web.HTTPPaymentRequired
+
+
+async def _verify_user_file(message: PendingStoreMessage, size: int, file_io) -> None:
+    file_io.seek(0)
+    content = file_io.read(size)
+    item_content = {}
+    if message.item_content:
+        item_content = json.loads(message.item_content)
+    actual_item_hash = sha256(content).hexdigest()
+    client_item_hash = item_content["item_hash"]
+    if len(content) > (1000 * MiB):
+        raise web.HTTPRequestEntityTooLarge(
+            actual_size=len(content), max_size=(1000 * MiB)
+        )
+    elif actual_item_hash != client_item_hash:
+        raise web.HTTPUnprocessableEntity()
+
+
+class StorageMetadata(pydantic.BaseModel):
+    message: PendingStoreMessage
+    file_size: int
+    sync: bool
+
+
+async def storage_add_file_with_message(
+    request: web.Request, session: DbSession, chain_service, storage_metadata, file_io
+):
+    config = get_config_from_request(request)
+    mq_queue = None
+
+    pending_store_message = PendingStoreMessage.parse_obj(storage_metadata.message)
+    pending_message_db = PendingMessageDb.from_obj(
+        obj=pending_store_message, reception_time=dt.datetime.now(), fetched=True
+    )
+    await _verify_message_signature(
+        pending_message=storage_metadata.message, chain_service=chain_service
+    )
+    await _verify_user_balance(
+        session=session,
+        pending_message_db=pending_message_db,
+        size=storage_metadata.file_size,
+    )
+    await _verify_user_file(
+        message=storage_metadata.message,
+        size=storage_metadata.file_size,
+        file_io=file_io,
+    )
+
+    if storage_metadata.sync:
+        mq_channel = await get_mq_channel_from_request(request, logger=logger)
+        mq_queue = await mq_make_aleph_message_topic_queue(
+            channel=mq_channel,
+            config=config,
+            routing_key=f"*.{pending_message_db.item_hash}",
+        )
+
+    session.add(pending_message_db)
+    session.commit()
+    if storage_metadata.sync and mq_queue:
+        mq_message = await mq_read_one_message(mq_queue, 30)
+
+        if mq_message is None:
+            raise web.HTTPAccepted()
+        if mq_message.routing_key is not None:
```

Comment:
What happens if `routing_key` is None?

Note that since you subscribe using a routing key, this case should never happen.